kpatch-build/create-diff-object.c

```diff
@@ -3754,114 +3753,68 @@ static void kpatch_create_callbacks_objname_rela(struct kpatch_elf *kelf, char *
 	}
 }
 
-/*
- * Allocate the mcount/patchable_function_entry sections which must be done
- * before the patched object is torn down so that the section flags can be
- * copied.
- */
-static void kpatch_alloc_mcount_sections(struct kpatch_elf *kelf, struct kpatch_elf *kelfout)
+static void kpatch_set_pfe_link(struct kpatch_elf *kelf)
 {
-	int nr;
-	struct symbol *sym;
-	int text_idx = 0;
+	struct section* sec;
+	struct rela *rela;
 
-	nr = 0;
-	list_for_each_entry(sym, &kelfout->symbols, list) {
-		if (sym->type == STT_FUNC && sym->status != SAME &&
-		    sym->has_func_profiling) {
-			text_idx = sym->sec->index;
-			nr++;
+	list_for_each_entry(sec, &kelf->sections, list) {
+		if (strcmp(sec->name, "__patchable_function_entries")) {
+			continue;
 		}
-	}
-
-	/* create text/rela section pair */
-	switch(kelf->arch) {
-	case AARCH64: {
-		struct section *sec;
-		int entries = multi_pfe ? 1 : nr;
-		int copies = multi_pfe ? nr : 1;
-		int flags = 0, rflags = 0;
 
-		/*
-		 * Depending on the compiler the __patchable_function_entries section
-		 * can be ordered or not, copy this flag to the section we created to
-		 * avoid:
-		 * ld: __patchable_function_entries has both ordered [...] and unordered [...] sections
-		 */
-		sec = find_section_by_name(&kelf->sections, "__patchable_function_entries");
-		if (sec) {
-			flags = (sec->sh.sh_flags & (SHF_LINK_ORDER|SHF_WRITE));
-			if (sec->rela)
-				rflags = (sec->rela->sh.sh_flags & (SHF_LINK_ORDER|SHF_WRITE));
+		if (!sec->rela) {
+			continue;
 		}
-
-		for (nr = 0; nr < copies; nr++) {
-			sec = create_section_pair(kelfout,
-						  "__patchable_function_entries",
-						  sizeof(void *), entries);
-
-			sec->sh.sh_flags |= flags;
-			if (sec->rela)
-				sec->rela->sh.sh_flags |= rflags;
-			if (multi_pfe)
-				sec->sh.sh_link = 0;
-			else
-				sec->sh.sh_link = text_idx;
+		list_for_each_entry(rela, &sec->rela->relas, list) {
+			rela->sym->sec->pfe = sec;
 		}
-		break;
-	}
-	case PPC64:
-	case X86_64:
-	case S390:
-		create_section_pair(kelfout, "__mcount_loc", sizeof(void *), nr);
-		break;
-	default:
-		ERROR("unsupported arch\n");
 	}
 }
 
 /*
- * Populate the mcount sections allocated by kpatch_alloc_mcount_sections()
- * previously.
  * This function basically reimplements the functionality of the Linux
  * recordmcount script, so that patched functions can be recognized by ftrace.
  *
  * TODO: Eventually we can modify recordmount so that it recognizes our bundled
  * sections as valid and does this work for us.
  */
-static void kpatch_populate_mcount_sections(struct kpatch_elf *kelf)
+static void kpatch_create_mcount_sections(struct kpatch_elf *kelf)
 {
 	int nr, index;
-	struct section *sec, *relasec;
+	struct section *relasec;
 	struct symbol *sym;
 	struct rela *rela, *mcount_rela;
 	void **funcs;
+	bool pfe_per_function;
```

Comment:
pfe_per_function should be default false.
Otherwise, it will be used without initialized when arch !=AARCH64

Reply:
Thanks for catch this. I was my mistake. But sorry, I do not have permission to push the "from branch". I will try to send another PR to swine's repo later.